adr/graph-feature-flags.md

```diff
@@ -0,0 +1,212 @@
+
+This will enable both the users-v2 and posts-v2 feature flags
+
+```shell
+curl -X POST https://graph.domain.com/graphql \
+  -H "Content-Type: application/json" \
+  -H "X-Feature-Flag: v2"
+  -d '{"query": "{ users { id name } posts { id title } }"}'
+```
+
+9. Delete the group feature flag and both feature flags
+
+```shell
+wgc ffg delete v2
+wgc ff delete users-v2
+wgc ff delete posts-v2
+```
+
+## Context
+
+### Feature Flags always replace the subgraph
+
+Feature Flags are always replacing the subgraph they are associated with.
+This means that a feature flag can only be associated with one subgraph at a time.
+If a feature flag is associated with a subgraph, it will replace the subgraph in the federated graph.
+
+### Conflicting Feature Flags that replace the same subgraph are allowed
+
+It's possible to have multiple feature flags that replace the same subgraph.
+This is useful and allowed for testing different variants of the same subgraph.
+
+### Feature Flags that replace the same subgraph are not allowed in a group feature flag
+
+A feature flag group is never allowed to contain multiple feature flags that replace the same subgraph.
+This would result in an ambiguous configuration.
+
+### Limiting a feature flag to specific federated graphs
+
+We've already established a pattern (label matching) to limit a subgraph to specific federated graphs.
+The same pattern can be used to limit a feature flag to specific federated graphs.
+
+```shell
+wgc federated-graph create production --label-matcher env=staging --routing-url https://graph.domain.com/graphql
+wgc ff create users-v2 --label env=staging --routing-url https://users-v2.domain.com
+wgc ff publish users-v2 --schema ./subgraph-users-v2/schema.graphql
+```
+
+### Combinations of feature flags are only possible via group feature flags
+
+In a previous version of Feature Flags, it was possible to automatically combine multiple feature flags.
+This was removed because it would have resulted in a exponential number of possible combinations,
+which would have required us to limit the total number of feature flags to a very low number.
+
+Instead, we require users to explicitly create group feature flags to combine multiple feature flags.
+With 20 active feature flags and 5 groups, we would have 20+5=25 possible combinations.
+If all 20 feature flags were combined, we would have 20*20=400 possible combinations.
+
+Each time a subgraph gets published, the Control Plane is required to check if all feature flags are still valid,
+which means that the number of possible combinations defines the composition check runs required.
+
+### Composition Checks need to be Feature Flag aware
+
+When a subgraph gets published, the Control Plane needs to check if the subgraph is compatible with all active feature flags.
+If errors occur, they need to be reported in such a way that the user can easily identify which feature flag caused the error.
```

Comment:
Will this be reported in the CLI and studio?

Reply:
yes, both